yesod-core/Yesod/Routes/TH/RenderRoute.hs

```diff
@@ -8,26 +8,30 @@ module Yesod.Routes.TH.RenderRoute
     ) where
 
 import Yesod.Routes.TH.Types
+#if MIN_VERSION_template_haskell(2,11,0)
+import Language.Haskell.TH (conT)
+#endif
 import Language.Haskell.TH.Syntax
 import Data.Maybe (maybeToList)
 import Control.Monad (replicateM)
 import Data.Text (pack)
 import Web.PathPieces (PathPiece (..), PathMultiPiece (..))
 import Yesod.Routes.Class
 #if __GLASGOW_HASKELL__ < 710
+import Control.Applicative ((<$>))
 import Data.Monoid (mconcat)
 #endif
 
 -- | Generate the constructors of a route data type.
-mkRouteCons :: [ResourceTree Type] -> ([Con], [Dec])
-mkRouteCons =
-    mconcat . map mkRouteCon
+mkRouteCons :: [ResourceTree Type] -> Q ([Con], [Dec])
+mkRouteCons rttypes =
+    mconcat <$> mapM mkRouteCon rttypes
```

Comment:
Now that I've had my morning coffee, I can give a more coherent explanation :).

As described in the GHC 8.0 mirgation guide ( https://ghc.haskell.org/trac/ghc/wiki/Migration/8.0 ), the `DataD` constructor has changed:

```
    - dataD :: CxtQ -> Name -> [TyVarBndr]               -> [ConQ] -> [Name] -> DecQ
    + dataD :: CxtQ -> Name -> [TyVarBndr] -> Maybe Kind -> [ConQ] -> CxtQ   -> DecQ
```

Specifiacally the last field has changed from `[Name]` to `CxtQ` which is just a type synonym for `Q Cxt`. The existing code for ghc < 8.0 already had value of type `[Name]` and the way to convert value `xs :: [Name]` to `CxtQ` is to use `mapM conT xs`.

Once we need the `Q` monad there, the whole function needs to be in the `Q` monad.